### PR TITLE
Create Smarthost for specific email or domain recipient

### PR DESCRIPTION
**Pull request by stephdl**
Create a smarthost to send email for a specific domain/email address when the default smtp is refused  by the remote smtp

https://github.com/NethServer/dev/issues/6169

### CODE REVIEW

**Comment by stephdl** [comment]
I think I am ready for reviews, my last [commit](https://github.com/NethServer/nethserver-mail/pull/198/commits/a104d3bd7ca0cdc5827606ceb904f75752e4206c) sort the file tls_policy where postfix store all informations relative on which port to contact a remote smtp and if it tries `may` or force `encrypt`. 
With the db smarthosts we can have several the same `Host` stored, this is right, but the old template expand for each DB key the Host here if it was enabled. This is wrong because postmap complained.

In short we have to sort the database, and use only one time `hostname:port    encrypt|may` in this file

File: smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender [hunk 1]
```diff
@@ -1,13 +1,34 @@
...
-        $OUT .= "[" . $_->prop('Host') . "]:" . $_->prop('Port') . "\t\t" . ($tls eq 'enabled' ? 'encrypt' : 'may') ."\n";
+        my $tls = ($_->prop('TlsStatus') eq 'enabled') ? 'tls' : '';
+        my $host = $_->prop('Host') || next;
+        my $port = $_->prop('Port')|| next;
+
+        if (! exists $hash{$host}->{'Port'}) {
+            $hash{$host}->{'Port'} = $port.$tls;
+        } elsif (!grep {$_ eq $port.$tls } split(',',$hash{$host}->{'Port'})) {
+            $hash{$host}->{'Port'} = $hash{$host}->{'Port'}.",$port$tls";
+          print $hash{$host}->{'Port'}."\n";
```
Comment (gsanchietti) [inline smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender:19]:
```suggestion
```

File: smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender [hunk 2]
```diff
@@ -1,13 +1,34 @@
...
+        }
     }
-}
 
+    foreach my $k (keys %hash) {
+        foreach my $v (split(',',$hash{$k}{'Port'})) {
+           # print $hash{$k}{'Port'}."\n";
```
Comment (gsanchietti) [inline smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender:25]:
```suggestion
```

File: smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender [hunk 3]
```diff
@@ -1,13 +1,34 @@
+
```
Comment (gsanchietti) [inline smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender:1]: is this new line really useful? :thinking:

File: common/etc/e-smith/templates/etc/postfix/transport/20smarthost_recipient [hunk 4]
```diff
@@ -0,0 +1,21 @@
+
```
Comment (gsanchietti) [inline common/etc/e-smith/templates/etc/postfix/transport/20smarthost_recipient:1]: Maybe this empty line can be removed?

File: smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender (hunk 1 above)
Comment (stephdl) [reply smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender:19]: oups...sorry.

Comment (stephdl) [reply smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender]: done

File: smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender (hunk 3 above)
Comment (stephdl) [reply smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender]: done

File: smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender (hunk 2 above)
Comment (stephdl) [reply smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender]: done

File: common/etc/e-smith/templates/etc/postfix/transport/20smarthost_recipient (hunk 4 above)
Comment (stephdl) [reply common/etc/e-smith/templates/etc/postfix/transport/20smarthost_recipient:1]: done

**Comment by stephdl** [comment]
> Could you also point me the line where the sort happens? I couldn't find it

L14 if the $hash{smtp.gmail.com}->{Port} doesn't exist then create it $hash{smtp.gmail.com}->{Port}=$port.$tls
L16 If the $port.$tls cannot be find in the array splitted $hash{$host}->{'Port'}, then add `",$port$tls"` to it

we can have 

```
smtp.gmail.com:587 encrypt
smtp.gmail.com:25 encrypt
```

but never 

```
smtp.gmail.com:587 encrypt
smtp.gmail.com:587 encrypt
```

**Comment by gsanchietti** [comment]
Thank you for the explanation, but it's not really a sort: the template is taking care to avoid duplicates

**Comment by stephdl** [comment]
yes you are right....remove duplicate

File: smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender [hunk 5]
```diff
@@ -4,10 +4,28 @@
...
+    foreach my $k (keys %hash) {
+        foreach my $v (split(',',$hash{$k}{'Port'})) {
+            if ($v =~ 'tls') {
+                $v =~ s|tls$||;
+                $OUT .= "[" . $k . "]:" . $v . "\t\t" ."encrypt\n";
+            } else {
+                $OUT .= "[" . $k . "]:" . $v . "\t\t" ."may\n";
+            }
+        }
+    }
```
Comment (gsanchietti) [inline smarthost/etc/e-smith/templates/etc/postfix/tls_policy/30smarthost-policy-sender:30]: Suggested change (replace from line 9 to 30):
```
        my $host = $_->prop('Host') || next;
        my $port = $_->prop('Port')|| next;
        my $k = ($_->prop('TlsStatus') eq 'enabled') ? "[$host]:$port\t\tencrypt" : "[$host]:$port\t\tmay";
        my $k_encrypt = "[$host]:$port\t\tencrypt";

        # replace duplicates: 'may' will replace 'encrypt'
        if ($k =~ m/may$/ && $hash{$k_encrypt}) {
            delete($hash{$k_encrypt});
        }
        $hash{$k} = 1;
    }

    foreach my $k (keys %hash) {
        $OUT .= "$k\n";
    }
```

Records with `may` encryption will override the `encrypt` ones.

**Comment by gsanchietti** [comment]
The same problem is present also in the `/etc/postfix/sasl_passwd` file.
Example:
```
[myserver.nethserver.org]:587   user1:pass1
[myserver.nethserver.org]:587   user2:pass2
```
Error:
```
[root@test tls_policy]# postmap /etc/postfix/sasl_passwd
postmap: warning: /etc/postfix/sasl_passwd.db: duplicate entry: "[myserver.nethserver.org]:587"
```

We need to add stronger validations inside the UI: the same host/port can't be used both on a sender or recipient. It's a postfix limitation and we must adhere to it.

File: common/etc/e-smith/templates/etc/postfix/transport/20smarthost_recipient (hunk 4 above)
Comment (stephdl) [reply common/etc/e-smith/templates/etc/postfix/transport/20smarthost_recipient:1]: done

File: smarthost/etc/e-smith/templates/etc/postfix/tls_policy/10smarthost-policy [hunk 6]
```diff
@@ -2,18 +2,35 @@
...
-    if($postfix{SmartHostStatus} ne 'enabled') {
-	return '';
+    use esmith::ConfigDB;
+    my $db = esmith::ConfigDB->open_ro('smarthosts');
+    my $dbc = esmith::ConfigDB->open_ro();
+
+    my $SmartHostName = $dbc->get_prop('postfix','SmartHostName') || '';
+    my $SmartHostPort = $dbc->get_prop('postfix','SmartHostPort') || '';
+    my $SmartHostTlsStatus = $dbc->get_prop('postfix','SmartHostTlsStatus') || '';
+    my $SmartHostStatus = $dbc->get_prop('postfix','SmartHostStatus') || 'disabled';
```
Comment (gsanchietti) [inline smarthost/etc/e-smith/templates/etc/postfix/tls_policy/10smarthost-policy:12]: The configuratio database is already accessible from templates, so why reopening it?
Example:
```
my $SmartHostName = $postfix{'SmartHostName'} || '';
```

Comment (stephdl) [reply smarthost/etc/e-smith/templates/etc/postfix/tls_policy/10smarthost-policy]: a bad habit I presume, it is  fixed

File: ui/public/i18n/language.json [hunk 7]
```diff
@@ -499,6 +502,8 @@
         "smarthost_not_found": "Smarthost not found",
         "must_start_with_at": "A domain must start with the '@' symbol like '@nethserver.org'",
         "valid_lessThan_1001": "Maximum allowed value is 1000",
-        "valid_lessThan_31": "Maximum allowed value is 30"
+        "valid_lessThan_31": "Maximum allowed value is 30",
+        "Host_already_in_use_with_another_account": "This host cannot have several accounts, you must use the account",
```
Comment (gsanchietti) [inline ui/public/i18n/language.json:506]:
```suggestion
        "Host_already_in_use_with_another_account": "This smart host has been already configured. If you want still to use it, you must enter again credentials for account ",
```
